ast/compile.go

```diff
@@ -3680,14 +3697,64 @@ func checkUnusedDeclaredVars(loc *Location, stack *localDeclaredVars, used VarSe
 	return errs
 }
 
-func rewriteSomeDeclStatement(g *localVarGenerator, stack *localDeclaredVars, decl *SomeDecl, errs Errors) Errors {
+func rewriteSomeDeclStatement(g *localVarGenerator, stack *localDeclaredVars, decl *SomeDecl, expr *Expr, errs Errors) (*Expr, Errors) {
 	for i := range decl.Symbols {
-		v := decl.Symbols[i].Value.(Var)
-		if _, err := rewriteDeclaredVar(g, stack, v, declaredVar); err != nil {
-			errs = append(errs, NewError(CompileErr, decl.Loc(), err.Error()))
+		switch v := decl.Symbols[i].Value.(type) {
+		case Var:
+			if _, err := rewriteDeclaredVar(g, stack, v, declaredVar); err != nil {
+				return nil, append(errs, NewError(CompileErr, decl.Loc(), err.Error()))
+			}
+		case Call:
+			switch len(v) {
+			case 3:
+				if !v[1].Value.IsGround() { // "some x in xs" declares "x" (via `member(x, xs)`)
+					e := expr.Copy()
+					e.Terms = []*Term{
+						RefTerm(VarTerm(Assign.Name)),
+						v[1].Copy(),
+						RefTerm(v[2].Copy(), NewTerm(g.Generate())),
+					}
+					return rewriteDeclaredAssignment(g, stack, e, errs)
+				}
+			case 4: // member/3
+				key, val := v[1], v[2]
+				e := expr.Copy()
+				if val.Value.IsGround() { // `some k, "v" in xs` ~> `xs[k0] = v` with fresh `k0`
+					if kv, ok := key.Value.(Var); ok {
+						gv, err := rewriteDeclaredVar(g, stack, kv, declaredVar)
+						if err != nil {
+							return nil, append(errs, NewError(CompileErr, decl.Loc(), err.Error()))
+						}
+						e.Terms = []*Term{
+							RefTerm(VarTerm(Equality.Name)),
+							val.Copy(),
+							RefTerm(v[3].Copy(), NewTerm(gv)),
+						}
+						return e, errs
+					}
+					if !key.Value.IsGround() { // unsupported
+						return nil, append(errs, NewError(CompileErr, decl.Loc(), "only var keys may be non-ground"))
+					}
+				} else { // val non-ground, rewrite to `v := xs[k]`, regardless of ground-ness of key
+					e.Terms = []*Term{
+						RefTerm(VarTerm(Assign.Name)),
+						val.Copy(),
+						RefTerm(v[3].Copy(), key.Copy()),
+					}
+				}
+				return rewriteDeclaredAssignment(g, stack, e, errs)
+			}
+
+			// If it's ground, we merely remove the "some" wrapping, so
+			//     some "foo" in ["foo", "bar"]
+			// becomes
+			//     "foo" in ["foo", "bar"]
```

Comment:
It's not mentioned explicitly, but that logic is the same for `member` and `member/3`.